### DISCUSSION TITLE
Rust in the kernel

### DISCUSSION THREAD

**Discussion by 0xdeadbad** [vvaltchev/tilck#153] [Q&A] [answered]
Is Rust allowed for the development of  some parts of the kernel?

**Comment by vvaltchev** [comment, marked as answer]
Hello! This is definitively an interesting question. 
When I start developing Tilck in 2016, Rust was barely just born and it didn't make any sense to use it. Now, everything is different and Rust is entering in the Linux kernel too. Therefore, Rust seems to be a valid language for low-level code. 

For Tilck, the biggest problem with Rust is the toolchain: I spent quite some time working on the `build_toolchain` script and on Tilck's CMake-based build system in order to allow switching easily between multiple pre-built GCC toolchains (from [toolchains.bootlin.com](https://toolchains.bootlin.com)), even across architectures. For Rust there is nothing _exactly_ like that yet. Linux doesn't care about such things because the whole toolchain configuration (e.g. cross-compilers etc.) is expected to happen outside of the project, while in Tilck everything is built-in, because it's easier for the developers this way. In other words, Tilck supports cross-compiling as part of the project itself.

Even if there was something like pre-compiled toolchains for each architecture for Rust, it will require an significant amount of time to integrate all of that in Tilck. Realistically, supporting Rust would mean adding an extra step to download and locally build the cross-compiler toolchain, in addition to all the rest (e.g. build-system integration, creation of Rust definitions of structures etc.). 

I don't plan to work on the Rust support any time soon, especially because I have a full-time job that keeps be super-busy. But, in the very long term (years) I might add support for Rust in the above-described way and probably start porting some of the more high-level code into it. I recognize that many younger developers will totally appreciate that.

**Reply by 0xdeadbad** [reply]
Thank you very for taking your time to write a detailed and comprehensive answer.

**Reply by vvaltchev** [reply]
You're welcome :-)
